keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+
+```yaml
+topologyHints:
+  zones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
+updated to filter endpoints based on topology hints when the following
+conditions are true:
+
+- Kube-Proxy is able to determine the zone it is running within (likely based
+  on node labels).
+- The `topology.kubernetes.io/topology-aware-routing` annotation is set to
+  `Auto` for the Service.
+- At least one endpoint for the Service has a hint pointing to the zone
+  Kube-Proxy is running within.
+- All endpoints for the Service have zone hints.
+
+When the above conditions are true, kube-proxy will only route traffic to
+endpoints with a hint referring to the zone Kube-Proxy is running within.
+
+### EndpointSlice Controller
+
+When the `TopologyAwareHints` feature gate is enabled and the
+`topology.kubernetes.io/topology-aware-routing` annotation is set to `Auto`
+for a Service, the EndpointSlice controller will add hints to EndpointSlices.
+These hints will indicate where an endpoint should be consumed by proxy
+implementations to enable topology aware routing.
+
+The EndpointSlice controller will determine how many endpoints should be
+available for each zone based on the proportion of CPU cores in each zone. If
+it is not possible to determine the number CPU cores, 1 core per node will be
+assumed for calculations.
+
+#### Example
+
+zone-a: 20 CPU cores
+zone-b: 16 CPU cores
+zone-c: 14 CPU cores
+
+In this scenario, the following proportion of endpoints would be allocated for
+each Service:
+
+zone-a: 40%
+zone-b: 32%
+zone-c: 28%
+
+When allocating endpoints to meet this distribution, keeping endpoints in the
+same zone will be prioritized. When same-zone endpoints are exhausted, endpoints
+will be taken from zones that have excess capacity.
+
+#### Overload
+
+Overload is a key concept for this proposal. This occurs when there are less
+endpoints for a zone than a perfect distribution would result in. For example,
+in a 3-zone cluster where each zone has an equivalent size, an EndpointSlice for
+a 4 endpoint service would not receive any zone hints. The expected number of
+endpoints per zone would be 1.33, and 2 of the 3 zones would only have 1
+endpoint allocated. This means that endpoints for these zones would be likely to
+receive 33% more traffic than a perfectly balanced scenario. In this case, the
+"Overload" for those zones would be 33%.
+
+Overload Threshold represents the maximum acceptable overload for this algorithm
+before changes are required. If the overload threshold is reached, the
+controller will attempt to redistribute endpoints to get below this threshold.
```

Comment:
I'm a bit lost here, how can and endpoint be redistributed, does it not belong to a zone?

Reply:
The key concept in this proposal is that each endpoint would include both the zone it is from (already does that) and a hint suggesting which zone should consume it. In most cases, those will be identical, but when endpoints are imbalanced, the hint may suggest that a different zone should consume the endpoint.

Reply:
Updated the summary to hopefully explain this better. Let me know if the KEP could still use some more clarification around this.

Reply:
got it now,   the `will attempt to redistribute endpoints` mean that `the controller will attempt to modify the endpoints hints so the endpoint distribution get below this threshold` , right?

Reply:
yes